felix/dataplane/linux/endpoint_mgr.go

```diff
@@ -1192,6 +1245,12 @@ func configureInterface(name string, ipVersion int, writeProcSys procSysWriter)
 		if err != nil {
 			return err
 		}
+		// Disable kernel rpf check for interfaces that have rpf filtering explicitely disabled
+		// This is set only in IPv4 mode as there's no equivalent sysctl in IPv6
+		err = writeProcSys(fmt.Sprintf("/proc/sys/net/ipv4/conf/%s/rp_filter", name), rpFilter)
```

Comment:
@mgleung is there anything we need to worry about here given the run-as-non-root work?

Reply:
Looks like we do something similar for egress gateways in calico-enterprise so this should be OK?

Reply:
I think that this should be fine if all we are looking for is support for running as non-root. Usually the issues around the sysctls comes from trying to run the containers as non-privileged.

> Looks like we do something similar for egress gateways in calico-enterprise so this should be OK?
Our non-privileged configuration does not yet support calico-enterprise features so it's unfortunately not a very good metric for if we can support it.

I'm just looking at these changes now so I might have this wrong, but from what I've covered so far, it seems that setting this is dependent on whether or not any source spoofing IPs are configured for an interface name. Since I think that the configuration is set through the pod annotations which would be after the pod/container has been created (and our CNI plugin has already been called), we wouldn't be able to easily add this call to the CNI plugin to make it work (https://github.com/projectcalico/calico/blob/master/cni-plugin/pkg/dataplane/linux/dataplane_linux.go#L381). Is there a way we could signal that the pod is configured for source IP spoofing on pod creation?

Reply:
The CNI can process the annotation if it is set at the pod creation time: https://github.com/projectcalico/calico/pull/5742/files#diff-c7a87df8a530a4bef96d896dc71262034477eb30fe4c2418d9ddfa89714d42eeR480
As you're saying, the drawback is that the change won't be taken into account if the annotation is added or removed afterwards. 

Another option would be to have the CNI always disable the kernel rpf check (`rp_filter=0`) on the pod interfaces, since the rpf check is also enforced in iptables anyway that should not change what the pods can do.

However, it looks like in unprivileged mode, calico still runs with CAP_NET_ADMIN (), which seems to allow setting this sysctl, so I don't think we should have an issue there. I tested by dropping every capability except net_admin using:
```
sudo capsh --drop=cap_chown,cap_dac_override,cap_dac_read_search,cap_fowner,cap_fsetid,cap_kill,cap_setgid,cap_setuid,cap_setpcap,cap_linux_immutable,cap_net_bind_service,cap_net_broadcast,cap_net_raw,cap_ipc_lock,cap_ipc_owner,cap_sys_module,cap_sys_rawio,cap_sys_chroot,cap_sys_ptrace,cap_sys_pacct,cap_sys_admin,cap_sys_boot,cap_sys_nice,cap_sys_resource,cap_sys_time,cap_sys_tty_config,cap_mknod,cap_lease,cap_audit_write,cap_audit_control,cap_setfcap,cap_mac_override,cap_mac_admin,cap_syslog,cap_wake_alarm,cap_block_suspend,cap_audit_read -- -c "capsh --print; sysctl -w net.ipv4.conf.ens5.rp_filter=0"
```

Reply:
I think that even with `CAP_NET_ADMIN` we should not be able to set sysctls when the pod is running as `privileged: false` since I don't think that `rp_filter` is in the safe set. I did test it a while back though so maybe that's changed.

> Another option would be to have the CNI always disable the kernel rpf check (rp_filter=0) on the pod interfaces, since the rpf check is also enforced in iptables anyway that should not change what the pods can do.

I know that `rp_filter` isn't strictly required because we have iptables checks as well but it makes me a little squeamish to remove it across all of our interfaces. Also this shouldn't block any of this code from going in since non-privileged Calico trades the access to all the newest features for the extra security. I would say at the very least, we should probably add a note to the docs (https://github.com/projectcalico/calico/blob/master/calico/security/non-privileged.md) to at least call out that it will not work with VPP, and we can follow up with disabling rpf checks across all interfaces from the CNI plugin later.

Reply:
Thanks @mgleung , I just added a note to `calico/security/non-privileged.md`.